## Reading CSV files with Encodings

In [1]:
import pandas as pd
laptops = pd.read_csv("/Users/brindhamanivannan/datasets/laptops.csv")
print(laptops.info())

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 4: invalid continuation byte

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 4: invalid continuation byte
    
'utf-8' is a type of encoding. Computers, at their lowest levels, can only understand binary - 0 and 1- and encodings are systems for representing characters in binary.  

The most common encodings are:

- UTF-8
- Latin-1 (also known as ISO-8859-1)
- Windows-1251

Since the pandas.read_csv() function already tried to read in the file with UTF-8 and failed, we know the file's not encoded with that format. Let us try Latin-1.

In [2]:
import pandas as pd
laptops = pd.read_csv("/Users/brindhamanivannan/datasets/laptops.csv", encoding="Latin-1")
print(laptops.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Manufacturer              1303 non-null   object
 1   Model Name                1303 non-null   object
 2   Category                  1303 non-null   object
 3   Screen Size               1303 non-null   object
 4   Screen                    1303 non-null   object
 5   CPU                       1303 non-null   object
 6   RAM                       1303 non-null   object
 7    Storage                  1303 non-null   object
 8   GPU                       1303 non-null   object
 9   Operating System          1303 non-null   object
 10  Operating System Version  1133 non-null   object
 11  Weight                    1303 non-null   object
 12  Price (Euros)             1303 non-null   object
dtypes: object(13)
memory usage: 132.5+ KB
None


## Observations

1. Every column is represented as Object type, indicating they are represnted by strings, not numbers.
2. Operating System Version column has null values
3. The upper case and lower case letters in the column names are going to be harder to work with
4. Storage column has a space infront of it
5. These quirks with column labels can sometimes be hard to spot, so removing extra whitespaces from all column names will save us more work in the long run.

In [3]:
print(laptops.columns)

Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', ' Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')


## Cleaning column names

In [4]:
# Removing white spaces in column names

new_columns = []
for c in laptops.columns:
    stripped_column = c.strip()
    new_columns.append(stripped_column)

print(new_columns)

['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen', 'CPU', 'RAM', 'Storage', 'GPU', 'Operating System', 'Operating System Version', 'Weight', 'Price (Euros)']


In [5]:
laptops.columns = new_columns
print(laptops.columns)

Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', 'Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')


Now, we have removed whitespaces from the column names.

Now, we will do the following

- Replacing spaces with underscores.
- Removing special characters.
- Making all labels lowercase.
- Shortening any long column names.

This way, the column labels will be all lower case with no special characters and that will make our life easier!

In [6]:
# Function that uses Python string methods to clean our column labels

import pandas as pd
laptops = pd.read_csv("/Users/brindhamanivannan/datasets/laptops.csv", encoding='Latin-1')

def clean_col(col):
    col = col.strip()
    col = col.replace("Operating System","os")
    col = col.replace(" ","_")
    col = col.replace("(","")
    col = col.replace(")","")
    col = col.lower()
    return col

# A loop to apply that function to each label

new_columns = []
for c in laptops.columns:
    clean_c = clean_col(c)
    new_columns.append(clean_c)

print(new_columns)

['manufacturer', 'model_name', 'category', 'screen_size', 'screen', 'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight', 'price_euros']


In [7]:
laptops.columns = new_columns 
print(laptops.columns)

Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight',
       'price_euros'],
      dtype='object')


## Converting String Columns to Numeric

In [8]:
laptops.head(10)

manufacturer       model_name   category screen_size  \
0        Apple      MacBook Pro  Ultrabook       13.3"   
1        Apple      Macbook Air  Ultrabook       13.3"   
2           HP           250 G6   Notebook       15.6"   
3        Apple      MacBook Pro  Ultrabook       15.4"   
4        Apple      MacBook Pro  Ultrabook       13.3"   
5         Acer         Aspire 3   Notebook       15.6"   
6        Apple      MacBook Pro  Ultrabook       15.4"   
7        Apple      Macbook Air  Ultrabook       13.3"   
8         Asus  ZenBook UX430UN  Ultrabook       14.0"   
9         Acer          Swift 3  Ultrabook       14.0"   

                               screen                         cpu   ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   
5                            1366x768     AMD A9-Series 9420 3GHz   4GB   
6  IPS Panel Retina Display 2880x1800        Intel Core i7 2.2GHz  16GB   
7                            1440x900        Intel Core i5 1.8GHz   8GB   
8                   Full HD 1920x1080  Intel Core i7 8550U 1.8GHz  16GB   
9         IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz   8GB   

               storage                           gpu       os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640    macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000    macOS        NaN   
2            256GB SSD         Intel HD Graphics 620    No OS        NaN   
3            512GB SSD            AMD Radeon Pro 455    macOS        NaN   
4            256GB SSD  Intel Iris Plus Graphics 650    macOS        NaN   
5            500GB HDD                 AMD Radeon R5  Windows         10   
6  256GB Flash Storage       Intel Iris Pro Graphics   Mac OS          X   
7  256GB Flash Storage        Intel HD Graphics 6000    macOS        NaN   
8            512GB SSD          Nvidia GeForce MX150  Windows         10   
9            256GB SSD        Intel UHD Graphics 620  Windows         10   

   weight price_euros  
0  1.37kg     1339,69  
1  1.34kg      898,94  
2  1.86kg      575,00  
3  1.83kg     2537,45  
4  1.37kg     1803,60  
5   2.1kg      400,00  
6  2.04kg     2139,97  
7  1.34kg     1158,70  
8   1.3kg     1495,00  
9   1.6kg      770,00

In [9]:
laptops.iloc[:10, 2:5]

category screen_size                              screen
0  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600
1  Ultrabook       13.3"                            1440x900
2   Notebook       15.6"                   Full HD 1920x1080
3  Ultrabook       15.4"  IPS Panel Retina Display 2880x1800
4  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600
5   Notebook       15.6"                            1366x768
6  Ultrabook       15.4"  IPS Panel Retina Display 2880x1800
7  Ultrabook       13.3"                            1440x900
8  Ultrabook       14.0"                   Full HD 1920x1080
9  Ultrabook       14.0"         IPS Panel Full HD 1920x1080

## Observation

Of these three columns, we have three different types of text data:

- category: Purely text data - there are no numeric values.
- screen_size: Numeric data stored as text data because of the " character.
- screen: A combination of pure text data with numeric data.


screen_size cannot be sort numerically as they are stored as text data. For example, if if we wanted to select laptops with screens 15" or larger, we'd be unable to do so.

## Data cleaning workflow

1. Explore the data in the column
2. Identify patterns and special cases
3. Remove non-digit characters
4. Convert the column to a numeric dtype
5. Rename column lable in required

#### Explore the data

In [10]:
print(laptops["screen_size"].dtype)
print(laptops["screen_size"].unique())

object
['13.3"' '15.6"' '15.4"' '14.0"' '12.0"' '11.6"' '17.3"' '10.1"' '13.5"'
 '12.5"' '13.0"' '18.4"' '13.9"' '12.3"' '17.0"' '15.0"' '14.1"' '11.3"']


In [11]:
print(laptops["ram"].dtype)
print(laptops["ram"].unique())

object
['8GB' '16GB' '4GB' '2GB' '12GB' '6GB' '32GB' '24GB' '64GB']


#### Identify patterns and special cases

Observations

- A series of digit and dot character followed by "
- No special cases
- Need to be converted to float
- ram column has GB at the end

#### Remove the non-digit characters

## Manipulate text data using vectorized string methods

Documentation: [Working with text data](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#method-summary)

Most vectorized string methods are available using the Series.str accessor, which means we can access them by adding str between the series name and the method name.

## str.replace()

In [12]:
laptops["screen_size"] = laptops["screen_size"].str.replace('"','')
print(laptops["screen_size"].unique())

['13.3' '15.6' '15.4' '14.0' '12.0' '11.6' '17.3' '10.1' '13.5' '12.5'
 '13.0' '18.4' '13.9' '12.3' '17.0' '15.0' '14.1' '11.3']


In [13]:
laptops["ram"] = laptops["ram"].str.replace('GB','')
unique_ram = laptops["ram"].unique()

In [14]:
print(unique_ram)

['8' '16' '4' '2' '12' '6' '32' '24' '64']


## Series.astype() 

Converting Columns to Numeric dtypes

In [15]:
laptops["screen_size"] = laptops["screen_size"].astype(float)
print(laptops["screen_size"].dtype)
print(laptops["screen_size"].unique())

float64
[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


In [16]:
laptops["ram"] = laptops["ram"].astype(int)
print(laptops["ram"].unique())

[ 8 16  4  2 12  6 32 24 64]


In [17]:
print(laptops.dtypes)

manufacturer     object
model_name       object
category         object
screen_size     float64
screen           object
cpu              object
ram               int64
storage          object
gpu              object
os               object
os_version       object
weight           object
price_euros      object
dtype: object


Now, we've converted our columns screen_size and ram to numeric dtypes.

## Renaming Columns

To stop us from losing information that helps us understand the data, we can use the DataFrame.rename() method to rename the column from screen_size to screen_size_inches and ram to ram_gb

## DataFrame.rename()

In [18]:
laptops.rename({"screen_size": "screen_size_inches"}, axis=1, inplace=True)
print(laptops.dtypes)

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram                     int64
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object


Above, we specify the axis=1 parameter so pandas knows that we want to rename labels in the column axis.

In [19]:
laptops.rename({"ram": "ram_gb"}, axis=1, inplace=True)


In [20]:
print(laptops.dtypes)

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram_gb                  int64
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object


## Descriptive statistics for a column
## Series.describe()


In [21]:
ram_gb_desc = laptops["ram_gb"].describe()

In [22]:
ram_gb_desc

count    1303.000000
mean        8.382195
std         5.084665
min         2.000000
25%         4.000000
50%         8.000000
75%         8.000000
max        64.000000
Name: ram_gb, dtype: float64

## Extracting values from Strings

In [23]:
print(laptops["gpu"].head())

0    Intel Iris Plus Graphics 640
1          Intel HD Graphics 6000
2           Intel HD Graphics 620
3              AMD Radeon Pro 455
4    Intel Iris Plus Graphics 650
Name: gpu, dtype: object


The information in this column seems to be a manufacturer (Intel, AMD) followed by a model name/number. Let's extract the manufacturer by itself so we can find the most common ones.

Because each manufacturer is followed by a whitespace character, we can use the Series.str.split() method to extract this data.

In [24]:
laptops["gpu_manufacturer"] = (laptops["gpu"]
                                       .str.split()
                                       .str[0]
                              )
                              
print(laptops["gpu_manufacturer"])

0       Intel
1       Intel
2       Intel
3         AMD
4       Intel
        ...  
1298    Intel
1299    Intel
1300    Intel
1301      AMD
1302    Intel
Name: gpu_manufacturer, Length: 1303, dtype: object


In [25]:
print(laptops["cpu"].head(10))

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
5       AMD A9-Series 9420 3GHz
6          Intel Core i7 2.2GHz
7          Intel Core i5 1.8GHz
8    Intel Core i7 8550U 1.8GHz
9    Intel Core i5 8250U 1.6GHz
Name: cpu, dtype: object


In [26]:
laptops["cpu_manufacturer"] = (laptops["cpu"]
                                       .str.split()
                                       .str[0]
                               )

laptops["cpu_manufacturer"]

0       Intel
1       Intel
2       Intel
3       Intel
4       Intel
        ...  
1298    Intel
1299    Intel
1300    Intel
1301    Intel
1302    Intel
Name: cpu_manufacturer, Length: 1303, dtype: object

## Series.value_counts()

In [27]:
cpu_manufacturer_counts = laptops["cpu_manufacturer"].value_counts()
cpu_manufacturer_counts

Intel      1240
AMD          62
Samsung       1
Name: cpu_manufacturer, dtype: int64

In [28]:
gpu_manufacturer_counts = laptops["gpu_manufacturer"].value_counts()
gpu_manufacturer_counts

Intel     722
Nvidia    400
AMD       180
ARM         1
Name: gpu_manufacturer, dtype: int64

## Correcting inconsistent values

In [29]:
print(laptops["os"].value_counts())

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          13
Mac OS          8
Android         2
Name: os, dtype: int64


Observations

- Mac OS and macOS. Two variations of the same operating system 


## Series.map()

The Series.map() method is ideal when we want to change multiple values in a column, but we'll use it now as an opportunity to learn how the method works.

In [30]:
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}

laptops["os"] = laptops["os"].map(mapping_dict)
print(laptops["os"].value_counts())

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          21
Android         2
Name: os, dtype: int64


One important thing to remember with Series.map() is that if a value from your series doesn't exist as a key in your dictionary, it will convert that value to NaN

## Dropping missing values

## DataFrame.isnull() 

In [31]:
print(laptops.isnull())

      manufacturer  model_name  category  screen_size_inches  screen    cpu  \
0            False       False     False               False   False  False   
1            False       False     False               False   False  False   
2            False       False     False               False   False  False   
3            False       False     False               False   False  False   
4            False       False     False               False   False  False   
...            ...         ...       ...                 ...     ...    ...   
1298         False       False     False               False   False  False   
1299         False       False     False               False   False  False   
1300         False       False     False               False   False  False   
1301         False       False     False               False   False  False   
1302         False       False     False               False   False  False   

      ram_gb  storage    gpu     os  os_version  we

In [32]:
print(laptops.isnull().sum())

manufacturer            0
model_name              0
category                0
screen_size_inches      0
screen                  0
cpu                     0
ram_gb                  0
storage                 0
gpu                     0
os                      0
os_version            170
weight                  0
price_euros             0
gpu_manufacturer        0
cpu_manufacturer        0
dtype: int64


We have only one column with null values, os_version, which has 170 missing values.

## How to handle missing values?



There are a few options for handling missing values:

1. Remove any rows that have missing values.
2. Remove any columns that have missing values.
3. Fill the missing values with some other value.
4. Leave the missing values as is.

The first two options are often used to prepare data for machine learning algorithms. Rows and columns with missing values cannot be used.

## DataFrame.dropna()

In [33]:
laptops_no_null_rows = laptops.dropna(axis=0)
laptops_no_null_cols = laptops.dropna(axis=1)

## Filling missing values

While dropping rows or columns is the easiest approach to deal with missing values, it may not always be the best approach. For example, removing a disproportionate amount of one manufacturer's laptops could change our analysis.

In [34]:
print(laptops["os_version"].value_counts(dropna=False))

10      1072
NaN      170
7         45
10 S       8
X          8
Name: os_version, dtype: int64


Because we set the dropna parameter to False, the result includes null values. 

In [35]:
os_with_null_v = laptops.loc[laptops["os_version"].isnull(),"os"]
print(os_with_null_v)


0       macOS
1       macOS
2       No OS
3       macOS
4       macOS
        ...  
1267    Linux
1277    Linux
1281    Linux
1291    Linux
1295    Linux
Name: os, Length: 170, dtype: object


In [36]:
print(os_with_null_v.value_counts())

No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: os, dtype: int64


## Observations

- The most frequent value is "No OS". This is important to note because if there is no os, there shouldn't be a version defined in the os_version column.
- Thirteen of the laptops that come with macOS do not specify the version. We can use our knowledge of MacOS to confirm that os_version should be equal to X.

In both of these cases, we can fill the missing values to make our data more correct. For the rest of the values, it's probably best to leave them as missing so we don't remove important values

In [37]:
value_counts_before = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
print(value_counts_before)

No OS        66
Linux        62
Chrome OS    27
macOS        13
Android       2
Name: os, dtype: int64


In [38]:
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X"
laptops.loc[laptops["os"] == "No OS", "os_version"] = "Version Unknown"

In [39]:
value_counts_after = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
print(value_counts_after)

Linux        62
Chrome OS    27
Android       2
Name: os, dtype: int64


## Cleaning a String column

In [40]:
print(laptops["weight"].head(20))

0     1.37kg
1     1.34kg
2     1.86kg
3     1.83kg
4     1.37kg
5      2.1kg
6     2.04kg
7     1.34kg
8      1.3kg
9      1.6kg
10    1.86kg
11    1.86kg
12    1.83kg
13     2.2kg
14    0.92kg
15    1.37kg
16     2.2kg
17    1.83kg
18     2.2kg
19    1.22kg
Name: weight, dtype: object


In [41]:
laptops["weight"] = laptops["weight"].str.replace("kgs","").str.replace("kg","").astype(float)

In [42]:
print(laptops["weight"].head())

0    1.37
1    1.34
2    1.86
3    1.83
4    1.37
Name: weight, dtype: float64


In [43]:
laptops.rename({"weight": "weight_kg"}, axis=1, inplace=True)


In [44]:
laptops.columns

Index(['manufacturer', 'model_name', 'category', 'screen_size_inches',
       'screen', 'cpu', 'ram_gb', 'storage', 'gpu', 'os', 'os_version',
       'weight_kg', 'price_euros', 'gpu_manufacturer', 'cpu_manufacturer'],
      dtype='object')

## More data cleaning

In [45]:
laptops["price_euros"].head()

0    1339,69
1     898,94
2     575,00
3    2537,45
4    1803,60
Name: price_euros, dtype: object

In [46]:
laptops["price_euros"] = laptops["price_euros"].str.replace(",","").astype(int)

In [47]:
laptops["price_euros"].head()

0    133969
1     89894
2     57500
3    253745
4    180360
Name: price_euros, dtype: int64

In [48]:
laptops["screen"].head()

0    IPS Panel Retina Display 2560x1600
1                              1440x900
2                     Full HD 1920x1080
3    IPS Panel Retina Display 2880x1800
4    IPS Panel Retina Display 2560x1600
Name: screen, dtype: object

In [49]:
laptops["screen_resolution"] = (laptops["screen"]
                                       .str.split()
                                       .str[-1]
                               )

laptops["screen_resolution"]

0       2560x1600
1        1440x900
2       1920x1080
3       2880x1800
4       2560x1600
          ...    
1298    1920x1080
1299    3200x1800
1300     1366x768
1301     1366x768
1302     1366x768
Name: screen_resolution, Length: 1303, dtype: object

In [50]:
laptops["cpu"].head()

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
Name: cpu, dtype: object

In [51]:
laptops["processor_speed"] = (laptops["cpu"]
                                       .str.split()
                                       .str[-1]
                               )

laptops["processor_speed"]

0       2.3GHz
1       1.8GHz
2       2.5GHz
3       2.7GHz
4       3.1GHz
         ...  
1298    2.5GHz
1299    2.5GHz
1300    1.6GHz
1301    2.5GHz
1302    1.6GHz
Name: processor_speed, Length: 1303, dtype: object

In [52]:
laptops.columns

Index(['manufacturer', 'model_name', 'category', 'screen_size_inches',
       'screen', 'cpu', 'ram_gb', 'storage', 'gpu', 'os', 'os_version',
       'weight_kg', 'price_euros', 'gpu_manufacturer', 'cpu_manufacturer',
       'screen_resolution', 'processor_speed'],
      dtype='object')

In [53]:
laptops.dtypes

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram_gb                  int64
storage                object
gpu                    object
os                     object
os_version             object
weight_kg             float64
price_euros             int64
gpu_manufacturer       object
cpu_manufacturer       object
screen_resolution      object
processor_speed        object
dtype: object

## DataFrame.to_csv()

To save the laptops dataframe to a CSV file laptops_cleaned.csv without index labels.

In [54]:
laptops.to_csv('/Users/brindhamanivannan/datasets/laptops_cleaned.csv',index=False)

## Business Questions

- Are laptops made by Apple more expensive than those made by other manufacturers?
- What is the best value laptop with a screen size of 15" or more?
- Which laptop has the most storage space?